sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,500 @@
+      abstract ConnectionConfiguration build();
+    }
+
+    public static ConnectionConfiguration create(List<String> hosts, String keyspace, int port) {
+      checkArgument(hosts != null, "ConnectionConfiguration.create(hosts, keyspace, port) "
+              + "called with null hosts");
+      checkArgument(hosts.size() >= 1, "ConnectionConfiguration.create(hosts, keyspace, port) "
+          + "called with an empty hosts list");
+      checkArgument(keyspace != null, "ConnectionConfiguration.create(hosts, keyspace, port) "
+          + "called with null keyspace");
+      checkArgument(port > 0, "ConnectionConfiguration.create(hosts, keyspace, port) called "
+          + "with invalid port number");
+      return new AutoValue_CassandraIO_ConnectionConfiguration.Builder()
+          .setHosts(hosts)
+          .setKeyspace(keyspace)
+          .setPort(port)
+          .build();
+    }
+
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("hosts", getHosts().toString()));
+      builder.addIfNotNull(DisplayData.item("keyspace", getKeyspace()));
+      builder.add(DisplayData.item("port", getPort()));
+    }
+
+    Cluster getCluster() {
+      LOG.debug("Connecting to Cassandra cluster");
+      return Cluster.builder().addContactPoints(getHosts().toArray(new String[0])).withPort
+          (getPort())
+          .build();
+    }
+  }
+
+  /**
+   * A {@link PTransform} to read data from Apache Cassandra. See {@link CassandraIO} for more
+   * information on usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    @Nullable abstract String getQuery();
+    @Nullable abstract String getTable();
+    @Nullable abstract String getRowKey();
+    @Nullable abstract Class<T> getEntityName();
+    @Nullable abstract Coder<T> getCoder();
+
+    abstract Builder<T> builder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setConnectionConfiguration(ConnectionConfiguration
+                                                         connectionConfiguration);
+      abstract Builder<T> setQuery(String query);
+      abstract Builder<T> setTable(String table);
+      abstract Builder<T> setRowKey(String rowKey);
+      abstract Builder<T> setEntityName(Class<T> entityName);
+      abstract Builder<T> setCoder(Coder<T> coder);
+      abstract Read<T> build();
+    }
+
+    public Read<T> withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "CassandraIO.read()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Read<T> withQuery(String query) {
```

Comment:
`Table` is used in the estimated size and split into bundles only.
`Query` is used in the reader to actually retrieve the data.
So, both have to be provided.